In [2]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [3]:
%sql mysql://USERNAME:PASSWORD@HOST/DATABASE

In [12]:
!pip install mysqlclient

  Using cached mysqlclient-2.1.1.tar.gz (88 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [16 lines of output]
      /bin/sh: mysql_config: command not found
      /bin/sh: mariadb_config: command not found
      /bin/sh: mysql_config: command not found
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/private/var/folders/nz/29g4kmmj6ljg5pf9j7mgf78c0000gn/T/pip-install-mrlz2cf5/mysqlclient_63ad9f3946f94b09bbbad176ac3fac75/setup.py", line 15, in <module>
          metadata, options = get_config()
        File "/private/var/folders/nz/29g4kmmj6ljg5pf9j7mgf78c0000gn/T/pip-install-mrlz2cf5/mysqlclient_63ad9f3946f94b09bbbad176ac3fac75/setup_posix.py", line 70, in get_config
          libs = mysql_config("libs")
        File "/private/var/folders/nz/29g4kmmj6ljg5pf

# EXPLORATORY SQL QUERY #1:

Purpose: To count the amount of songs per year, in order to double check that there is indeed 100 unique top songs from each year from 1970 through 2022 in the database.

In [5]:
%%sql

WITH top_songs AS (
	SELECT
		t.track_id,
		t.chart_year,
		t.track_name,
		t.artist,
		t.album,
		td.track_key,
		td.tempo,
		td.duration_ms,
		td.time_signature,
		ta.danceability,
		ta.energy,
		ta.loudness,
		ta.speechiness,
		ta.acousticness,
		ta.instrumentalness,
		ta.liveness,
		ta.valence
	FROM track t
	JOIN track_details td 
		ON t.track_id = td.track_id AND t.chart_year = td.chart_year
	JOIN track_attributes ta
		ON t.track_id = ta.track_id AND t.chart_year = ta.chart_year
	ORDER BY t.chart_year
)
SELECT
	chart_year,
	COUNT(DISTINCT(track_id))
FROM track
GROUP BY chart_year
ORDER BY chart_year;


 * mysql://admin:***@isba-dev-01.ciyo3znowdxs.us-east-1.rds.amazonaws.com/final_project
53 rows affected.


[(1970, 100),
 (1971, 100),
 (1972, 100),
 (1973, 100),
 (1974, 100),
 (1975, 100),
 (1976, 100),
 (1977, 100),
 (1978, 100),
 (1979, 100),
 (1980, 100),
 (1981, 100),
 (1982, 100),
 (1983, 100),
 (1984, 100),
 (1985, 100),
 (1986, 100),
 (1987, 100),
 (1988, 100),
 (1989, 100),
 (1990, 100),
 (1991, 100),
 (1992, 100),
 (1993, 100),
 (1994, 100),
 (1995, 100),
 (1996, 100),
 (1997, 100),
 (1998, 100),
 (1999, 100),
 (2000, 100),
 (2001, 100),
 (2002, 100),
 (2003, 100),
 (2004, 100),
 (2005, 100),
 (2006, 100),
 (2007, 100),
 (2008, 100),
 (2009, 100),
 (2010, 100),
 (2011, 100),
 (2012, 100),
 (2013, 100),
 (2014, 100),
 (2015, 100),
 (2016, 100),
 (2017, 100),
 (2018, 100),
 (2019, 100),
 (2020, 100),
 (2021, 100),
 (2022, 100)]

Discovery: I discovered that there was indeed 100 unique songs per year in the database, as expected.

Insight: This confirms that data entry was done correctly and that the database can be used for accurate queries.

# EXPLORATORY SQL QUERY #2:

Purpose: To see how many songs were top hits for more than 1 year, if any, and how many years they were a top hit.

In [10]:
%%sql

WITH top_songs AS (
	SELECT
		t.track_id,
		t.chart_year,
		t.track_name,
		t.artist,
		t.album,
		td.track_key,
		td.tempo,
		td.duration_ms,
		td.time_signature,
		ta.danceability,
		ta.energy,
		ta.loudness,
		ta.speechiness,
		ta.acousticness,
		ta.instrumentalness,
		ta.liveness,
		ta.valence
	FROM track t
	JOIN track_details td 
		ON t.track_id = td.track_id AND t.chart_year = td.chart_year
	JOIN track_attributes ta
		ON t.track_id = ta.track_id AND t.chart_year = ta.chart_year
	ORDER BY t.chart_year
)
SELECT
	track_id,
	track_name,
	artist,
	COUNT(chart_year) AS years_as_top_hit
FROM top_songs
GROUP BY track_id
HAVING COUNT(chart_year) > 1
ORDER BY COUNT(chart_year) DESC;


 * mysql://admin:***@isba-dev-01.ciyo3znowdxs.us-east-1.rds.amazonaws.com/final_project
125 rows affected.


[('1gsnWXeNlbfoHpaGRi4oQ0', 'Two Can Play That Game - Radio Edit', 'Bobby Brown', 3),
 ('6lanRgr6wXibZr8KgzXxBl', 'A Thousand Years', 'Christina Perri', 3),
 ('7tFiyTwD0nx5a1eklYtX2J', 'Bohemian Rhapsody - Remastered 2011', 'Queen', 3),
 ('01cQPNCptd662nikuY42zw', 'Have You Seen Her', 'The Chi-Lites', 2),
 ('09CtPGIpYB4BrO8qb1RGsF', 'Sorry', 'Justin Bieber', 2),
 ('0AcLrSfAEBQcUnHOTm5pXg', 'Get It On Tonite', 'Montell Jordan', 2),
 ('0BCPKOYdS2jbQ8iyB56Zns', 'Clocks', 'Coldplay', 2),
 ('0e8nrvls4Qqv5Rfa2UhqmO', 'THATS WHAT I WANT', 'Lil Nas X', 2),
 ('0fYVliAYKHuPmECRs1pbRf', 'Renegades', 'X Ambassadors', 2),
 ('0ntQJM78wzOLVeCUAW7Y45', 'Sex on Fire', 'Kings of Leon', 2),
 ('0obBFrPYkSoBJbvHfUIhkv', 'Sexy And I Know It', 'LMFAO', 2),
 ('0rFOs9paloAvEtzwDX1Kmc', 'Bubbly', 'Colbie Caillat', 2),
 ('0RiRZpuVRbi7oqRdSMwhQY', 'Sunflower - Spider-Man: Into the Spider-Verse', 'Post Malone', 2),
 ('0SiywuOBRcynK0uKGWdCnn', 'Bad Romance', 'Lady Gaga', 2),
 ('0t7fVeEJxO2Xi4H2K5Svc9', 'Send My Love (To Your New Lover)', 'Adele', 2),
 ('0U10zFw4GlBacOy9VDGfGL', 'We Found Love', 'Rihanna', 2),
 ('0u2P5u6lvoDfwTYjAADbn4', 'lovely (with Khalid)', 'Billie Eilish', 2),
 ('0UzsDmdpw0Q14KU4hieQss', 'Fill Me In', 'Craig David', 2),
 ('0W4Kpfp1w2xkY3PrV714B7', 'Ho Hey', 'The Lumineers', 2),
 ('0wwPcA6wtMf6HUMpIRdeP7', 'Hotline Bling', 'Drake', 2),
 ('0YImOCkIJ2PWhCXaURCZnY', 'I Know What You Want (feat. Flipmode Squad)', 'Busta Rhymes', 2),
 ('12dU3vAh6AFoJkisorfoUl', "Don't Look Back In Anger - Remastered", 'Oasis', 2),
 ('1DunhgeZSEgWiIYbHqXl0c', 'Latch', 'Disclosure', 2),
 ('1EAgPzRbK9YmdOESSMUm6P', 'Home', 'Phillip Phillips', 2),
 ('1IAzD1muglOxOcPbUHs70R', 'Whatever You Like', 'T.I.', 2),
 ('1j8z4TTjJ1YOdoFEDwJTQa', "Ain't It Fun", 'Paramore', 2),
 ('1jF7IL57ayN4Ity3jQqGu0', 'Try', 'P!nk', 2),
 ('1kxeWHF9PrCVZHvVskv8lg', 'Love Lockdown', 'Kanye West', 2),
 ('1M2nd8jNUkkwrc1dgBPTJz', 'Angels', 'Robbie Williams', 2),
 ('1nRwyxNsqCLeA17qR8Nfxx', 'Quit Playing Games (With My Heart)', 'Backstreet Boys', 2),
 ('1OxcIUqVmVYxT6427tbhDW', 'Fly Away', 'Lenny Kravitz', 2),
 ('21jGcNKet2qwijlDFuPiPb', 'Circles', 'Post Malone', 2),
 ('25IRjWUZ60n7KXxcB4DiX9', 'Strong Enough', 'Sheryl Crow', 2),
 ('27GmP9AWRs744SzKcpJsTZ', 'Jumpman', 'Drake', 2),
 ('28IEbk5a7twNTbUEvWslUb', 'Paralyzer', 'Finger Eleven', 2),
 ('29d0nY7TzCoi22XBqDQkiP', 'Running Up That Hill (A Deal With God) - 2018 Remaster', 'Kate Bush', 2),
 ('2ajhenDRiiAnCHHjsIOHhJ', 'Laffy Taffy', 'D4L', 2),
 ('2fOxxLz9JVe53itq6pfBak', "It's My Life", 'No Doubt', 2),
 ('2Fxmhks0bxGSBdJ92vM42m', 'bad guy', 'Billie Eilish', 2),
 ('2NBQmPrOEEjA8VbeWOQGxO', "Drop It Like It's Hot", 'Snoop Dogg', 2),
 ('2ouURa1AIXp3AvkS52Jry5', 'Every Morning', 'Sugar Ray', 2),
 ('2tnVG71enUj33Ic2nFN6kZ', 'Ride It', 'Regard', 2),
 ('2xLMifQCjDGFmkHkpNLD9h', 'SICKO MODE', 'Travis Scott', 2),
 ('34gCuhDGsG4bRPIf9bb02f', 'Thinking out Loud', 'Ed Sheeran', 2),
 ('35ItUJlMtjOQW3SSiTCrrw', 'Crazy Little Thing Called Love - Remastered 2011', 'Queen', 2),
 ('3G6hD9B2ZHOsgf4WfNu7X1', 'Team', 'Lorde', 2),
 ('3jagJCUbdqhDSPuxP8cAqF', 'Work It', 'Missy Elliott', 2),
 ('3jlbL2OTD5YmIunYzgQTAN', 'Live Your Life', 'T.I.', 2),
 ('3N4DI1vuTSX1tz7fa2NQZw', 'Rehab', 'Amy Winehouse', 2),
 ('3pXF1nA74528Edde4of9CC', "Don't", 'Bryson Tiller', 2),
 ('3tSuRiNgJmMqmYR23dCqnR', 'I Hate This Part', 'The Pussycat Dolls', 2),
 ('3w3y8KPTfNeOKPiqUTakBh', 'Locked out of Heaven', 'Bruno Mars', 2),
 ('3wKKkGVn7cVSfgPeYYIiXG', 'Kiss Kiss (feat. T-Pain)', 'Chris Brown', 2),
 ('3xrn9i8zhNZsTtcoWgQEAd', 'Since U Been Gone', 'Kelly Clarkson', 2),
 ('3yDhZq8f17SmumVmEyCaRN', 'Give It Up', 'KC & The Sunshine Band', 2),
 ('3yUcJwYu7fXAfqMj9krY6l', 'Thank You', 'Dido', 2),
 ('463CkQjx2Zk1yXoBuierM9', 'Levitating (feat. DaBaby)', 'Dua Lipa', 2),
 ('46n2EGFnPC3tzWCN1Aqe26', 'This I Promise You', '*NSYNC', 2),
 ('4AE7Lj39VnSZNOmGH2iZaq', 'Do For Love', '2Pac', 2),
 ('4BggEwLhGfrbrl7JBhC8EC', 'Butterfly', 'Crazy Town', 2),
 ('4eHbdreAnSOrDDsFfc4Fpm', 'I Will Always Love 

Discovery: 125 songs were top hits for more than 1 year, with 3 songs being a top hit for 3 separate years, and the rest of the songs returned by this query being a top hit for 2 years.

Insight: This shows that there is a fair amount of songs with enough lasting potential to be a top hit in separate years.  It also means that the DISTINCT() function will need to be used to avoid returning the same song under several years, if that is not wanted in a query.

# EXPLORATORY SQL QUERY #3:

Purpose: To see if there are different top hits with the same song title, and if so, how many?

In [11]:
%%sql

WITH top_songs AS (
	SELECT
		t.track_id,
		t.chart_year,
		t.track_name,
		t.artist,
		t.album,
		td.track_key,
		td.tempo,
		td.duration_ms,
		td.time_signature,
		ta.danceability,
		ta.energy,
		ta.loudness,
		ta.speechiness,
		ta.acousticness,
		ta.instrumentalness,
		ta.liveness,
		ta.valence
	FROM track t
	JOIN track_details td 
		ON t.track_id = td.track_id AND t.chart_year = td.chart_year
	JOIN track_attributes ta
		ON t.track_id = ta.track_id AND t.chart_year = ta.chart_year
	ORDER BY t.chart_year
)
SELECT
	track_name,
	COUNT(DISTINCT(track_id)) AS Num_of_Songs_with_Name
FROM top_songs
GROUP BY track_name
HAVING COUNT(DISTINCT(track_id)) > 1
ORDER BY COUNT(DISTINCT(track_id)) DESC;

 * mysql://admin:***@isba-dev-01.ciyo3znowdxs.us-east-1.rds.amazonaws.com/final_project
221 rows affected.


[('Angel', 5),
 ('Hold On', 4),
 ('Human', 4),
 ('Photograph', 4),
 ('Someday', 4),
 ('The Power Of Love', 4),
 ("When You're Gone", 4),
 ('Alone', 3),
 ('Always', 3),
 ('Around the World', 3),
 ('Beautiful', 3),
 ('Closer', 3),
 ('Crazy', 3),
 ('Dynamite', 3),
 ('Fire', 3),
 ('Forever', 3),
 ('Good Life', 3),
 ('Holiday', 3),
 ('Home', 3),
 ('How Deep Is Your Love', 3),
 ('I Like It', 3),
 ('In Your Eyes', 3),
 ("It's My Life", 3),
 ('Lady', 3),
 ('Life Is A Highway', 3),
 ('Mercy', 3),
 ('Missing You', 3),
 ('Money', 3),
 ('Oh Girl', 3),
 ('Perfect', 3),
 ('Ready Or Not', 3),
 ('Sing', 3),
 ('Sorry', 3),
 ('Stay', 3),
 ('Superstar', 3),
 ('Wishing Well', 3),
 ('A Moment Like This', 2),
 ('Again', 2),
 ("Ain't Nobody", 2),
 ('All By Myself', 2),
 ('All For You', 2),
 ('All I Want', 2),
 ('All This Time', 2),
 ('All Time Low', 2),
 ('American Woman', 2),
 ('Angel of Mine', 2),
 ('Angels', 2),
 ('Animal', 2),
 ('As', 2),
 ('As Long As You Love Me', 2),
 ('Axel F', 2),
 ('Baby', 2),
 ('Baby Come Back', 2),
 ('Baker Street', 2),
 ('Because of You', 2),
 ("Beggin'", 2),
 ('Behind Blue Eyes', 2),
 ('Big Yellow Taxi', 2),
 ('Blinding Lights', 2),
 ('Body', 2),
 ('Boyfriend', 2),
 ('Breathe', 2),
 ('Bridge over Troubled Water', 2),
 ('Brother Louie', 2),
 ('Burn', 2),
 ("C'est La Vie", 2),
 ('Call Me', 2),
 ("Can't Stop", 2),
 ("Can't You See", 2),
 ('Candy', 2),
 ('Cherish', 2),
 ('Cherry Bomb', 2),
 ('Cold Heart - PNAU Remix', 2),
 ('Come Undone', 2),
 ('Cruel Summer', 2),
 ('Cum on Feel the Noize', 2),
 ("Da Ya Think I'm Sexy?", 2),
 ('DÁKITI', 2),
 ('Dance Monkey', 2),
 ('Dangerous', 2),
 ('Do It Again', 2),
 ("Doin' Time", 2),
 ("Don't", 2),
 ("Don't Let The Sun Go Down On Me", 2),
 ("Don't Stop The Music", 2),
 ("Don't Turn Around", 2),
 ('Drive', 2),
 ('Easy On Me', 2),
 ('emotions', 2),
 ('Everything', 2),
 ('Everything I Own', 2),
 ('Express Yourself', 2),
 ('Faded', 2),
 ('Fantasy', 2),
 ('Fast Car', 2),
 ('Free', 2),
 ('Get Ready', 2),
 ('good 4 u', 2),
 ('Good Enough', 2),
 ('Happier', 2),
 ('Have You Ever', 2),
 ('Have You Seen Her', 2),
 ('Heartbreaker', 2),
 ('Heat Waves', 2),
 ('Heaven', 2),
 ('Hello', 2),
 ('Here', 2),
 ('Here With Me', 2),
 ('Hero', 2),
 ('High Hopes', 2),
 ('Higher', 2),
 ('Higher Ground', 2),
 ('Higher Love', 2),
 ('How Long', 2),
 ('Hurts So Good', 2),
 ('I Drove All Night', 2),
 ('I Miss You', 2),
 ('I Shot The Sheriff', 2),
 ('I Wish', 2),
 ('If', 2),
 ('Impossible', 2),
 ('In the Summertime', 2),
 ('Incomplete', 2),
 ('INDUSTRY BABY (feat. Jack Harlow)', 2),
 ("It's Raining Men", 2),
 ('Just the Way You Are', 2),
 ("Keep On Movin'", 2),
 ('Killing Me Softly With His Song', 2),
 ('Landslide', 2),
 ('Let It Go', 2),
 ('Let Me Love You', 2),
 ('Live And Let Die', 2),
 ('Lost Without You', 2),
 ('Love', 2),
 ('Love Machine', 2),
 ('Love Me For A Reason', 2),
 ('Love Song', 2),
 ('Love Train', 2),
 ('Mad World', 2),
 ('Magic', 2),
 ('Malibu', 2),
 ('Mandy', 2),
 ('Maneater', 2),
 ('Mine', 2),
 ('Miss Independent', 2),
 ('Mockingbird', 2),
 ('Mood (feat. iann dior)', 2),
 ('Motivation', 2),
 ('My Life', 2),
 ('My Way', 2),
 ('Never Can Say Goodbye', 2),
 ('No Hands (feat. Roscoe Dash & Wale)', 2),
 ('No Limit', 2),
 ('Now That We Found Love', 2),
 ('One', 2),
 ('One Call Away', 2),
 ('Only You', 2),
 ('Open Your Heart', 2),
 ('Physical', 2),
 ('Play That Funky Music', 2),
 ("Please Don't Go", 2),
 ('Poison', 2),
 ('positions', 2),
 ('Promise', 2),
 ('Put Your Records On', 2),
 ('Release Me', 2),
 ('Rich Girl', 2),
 ('Ride', 2),
 ('Rise', 2),
 ('Rock The Boat', 2),
 ('Roses', 2),
 ('Sacrifice', 2),
 ('Sailing', 2),
 ('Scream', 2),
 ('Sealed with a Kiss', 2),
 ('Sexy Eyes', 2),
 ('Shape of My Heart', 2),
 ('Shining Star', 2),
 ('Shivers', 2),
 ('Show Me The Way', 2),
 ('Skinny Love', 2),
 ('Slow Motion', 2),
 ('Smile', 2),
 ('Somebody', 2),
 ('Sorry Seems To Be The Hardest Word', 2),
 ('SOS', 2),
 ('Spotlight', 2),
 ('STAY (with Justin Bieber)', 2),
 ('Stay With Me', 2),
 ('Step by Step', 2),
 ('Stop'

Discovery: There are 221 song titles that are shared between songs.  The most instances of a shared title is "Angel" with 5 unique songs having that title.

Insight: This shows that there is a fair amount of top hit songs that share a title with other top hits.  Knowing this, it should not be a worry if an artist has songs with the same title as other top hits, as that won't prevent the song from being a top hit itself.

# EXPLORATORY SQL QUERY #4:

Purpose: To see who the top 100 artists are with the most top hits, and how many top hits did they have?

In [15]:
%%sql

WITH top_songs AS (
	SELECT
		t.track_id,
		t.chart_year,
		t.track_name,
		t.artist,
		t.album,
		td.track_key,
		td.tempo,
		td.duration_ms,
		td.time_signature,
		ta.danceability,
		ta.energy,
		ta.loudness,
		ta.speechiness,
		ta.acousticness,
		ta.instrumentalness,
		ta.liveness,
		ta.valence
	FROM track t
	JOIN track_details td 
		ON t.track_id = td.track_id AND t.chart_year = td.chart_year
	JOIN track_attributes ta
		ON t.track_id = ta.track_id AND t.chart_year = ta.chart_year
	ORDER BY t.chart_year
)
SELECT
	artist,
	COUNT(DISTINCT(track_id)) AS num_top_songs
FROM top_songs
GROUP BY artist
ORDER BY num_top_songs DESC
LIMIT 100;

 * mysql://admin:***@isba-dev-01.ciyo3znowdxs.us-east-1.rds.amazonaws.com/final_project
100 rows affected.


[('Various Artists', 93),
 ('Elton John', 31),
 ('Taylor Swift', 31),
 ('Michael Jackson', 30),
 ('Drake', 29),
 ('Madonna', 28),
 ('Rihanna', 26),
 ('Eminem', 24),
 ('U2', 22),
 ('Beyoncé', 22),
 ('Ariana Grande', 21),
 ('Britney Spears', 21),
 ('ABBA', 21),
 ('Calvin Harris', 20),
 ('Mariah Carey', 19),
 ('Kanye West', 19),
 ('Queen', 19),
 ('Whitney Houston', 18),
 ('Rod Stewart', 18),
 ('Usher', 18),
 ('Justin Bieber', 18),
 ('Ed Sheeran', 18),
 ('Janet Jackson', 18),
 ('Stevie Wonder', 18),
 ('P!nk', 18),
 ('Coldplay', 16),
 ('2Pac', 16),
 ('Chris Brown', 16),
 ('Black Eyed Peas', 16),
 ('George Michael', 16),
 ('Katy Perry', 15),
 ('The Weeknd', 15),
 ('Bon Jovi', 15),
 ('Bryan Adams', 14),
 ('Billy Joel', 14),
 ('Maroon 5', 14),
 ('JAY-Z', 14),
 ('Bruno Mars', 13),
 ('Lady Gaga', 13),
 ("Guns N' Roses", 13),
 ('Shawn Mendes', 13),
 ('Diana Ross', 13),
 ('Adele', 13),
 ('Bee Gees', 13),
 ('Post Malone', 13),
 ('Kendrick Lamar', 12),
 ('Westlife', 12),
 ('David Bowie', 12),
 ('Cher', 12),
 ('Billie Eilish', 12),
 ('Aerosmith', 12),
 ('Red Hot Chili Peppers', 12),
 ('Carpenters', 12),
 ('Fleetwood Mac', 12),
 ('Prince', 12),
 ('Bob Marley & The Wailers', 11),
 ('Def Leppard', 11),
 ("Destiny's Child", 11),
 ('Justin Timberlake', 11),
 ('Genesis', 11),
 ('Céline Dion', 11),
 ('Jennifer Lopez', 11),
 ('Avril Lavigne', 11),
 ('Jason Derulo', 11),
 ('Phil Collins', 11),
 ('One Direction', 11),
 ('Bad Bunny', 11),
 ('Christina Aguilera', 10),
 ('Lil Nas X', 10),
 ('UB40', 10),
 ('Lionel Richie', 10),
 ('Robbie Williams', 10),
 ('Donna Summer', 10),
 ('Mary J. Blige', 10),
 ('The Police', 10),
 ('Eagles', 10),
 ('Avicii', 10),
 ('Flo Rida', 10),
 ('Electric Light Orchestra', 10),
 ('Sting', 10),
 ('Sam Smith', 9),
 ('Selena Gomez', 9),
 ('Roxette', 9),
 ('Pitbull', 9),
 ('Imagine Dragons', 9),
 ('Wings', 9),
 ('Daryl Hall & John Oates', 9),
 ('David Guetta', 9),
 ('Bob Seger', 9),
 ('LL COOL J', 9),
 ('Marshmello', 9),
 ('The Rolling Stones', 9),
 ('Miley Cyrus', 9),
 ('Nelly', 9),
 ('Nirvana', 9),
 ('Take That', 9),
 ('Huey Lewis & The News', 9),
 ('INXS', 9),
 ('Nickelback', 9),
 ('Alicia Keys', 9)]

Discovery: The top 2 artists both have 31 distinct top hit songs.  All of the top 100 artists with the most hits have 9 or more hits.

Insight: This shows that there are a lot of artists with a considerable amount of top hits.  As such, we now know that artists who have had a top hit in the past have a high potential of having more, as it is not uncommon at all to have many top hits.

# EXPLORATORY SQL QUERY #5:

Purpose: To see how song durations have changed each decade.

In [16]:
%%sql

WITH top_songs AS (
	SELECT
		t.track_id,
		t.chart_year,
		t.track_name,
		t.artist,
		t.album,
		td.track_key,
		td.tempo,
		td.duration_ms,
		td.time_signature,
		ta.danceability,
		ta.energy,
		ta.loudness,
		ta.speechiness,
		ta.acousticness,
		ta.instrumentalness,
		ta.liveness,
		ta.valence
	FROM track t
	JOIN track_details td 
		ON t.track_id = td.track_id AND t.chart_year = td.chart_year
	JOIN track_attributes ta
		ON t.track_id = ta.track_id AND t.chart_year = ta.chart_year
	ORDER BY t.chart_year
)
SELECT
    FLOOR(chart_year / 10) * 10 AS decade,
    AVG(duration_ms) AS avg_duration_ms,
    MIN(duration_ms) AS min_duration_ms,
    MAX(duration_ms) AS max_duration_ms
FROM top_songs
GROUP BY FLOOR(chart_year / 10) * 10
ORDER BY decade;WITH top_songs AS (
	SELECT
		t.track_id,
		t.chart_year,
		t.track_name,
		t.artist,
		t.album,
		td.track_key,
		td.tempo,
		td.duration_ms,
		td.time_signature,
		ta.danceability,
		ta.energy,
		ta.loudness,
		ta.speechiness,
		ta.acousticness,
		ta.instrumentalness,
		ta.liveness,
		ta.valence
	FROM track t
	JOIN track_details td 
		ON t.track_id = td.track_id AND t.chart_year = td.chart_year
	JOIN track_attributes ta
		ON t.track_id = ta.track_id AND t.chart_year = ta.chart_year
	ORDER BY t.chart_year
)
SELECT
    FLOOR(chart_year / 10) * 10 AS decade,
    AVG(duration_ms) AS avg_duration_ms,
    MIN(duration_ms) AS min_duration_ms,
    MAX(duration_ms) AS max_duration_ms
FROM top_songs
GROUP BY FLOOR(chart_year / 10) * 10
ORDER BY decade;

 * mysql://admin:***@isba-dev-01.ciyo3znowdxs.us-east-1.rds.amazonaws.com/final_project
6 rows affected.
6 rows affected.


[(1970, Decimal('239874.0250'), 92867, 1561133),
 (1980, Decimal('259409.5730'), 126440, 820907),
 (1990, Decimal('261809.5876'), 121160, 718600),
 (2000, Decimal('238784.9660'), 127920, 448573),
 (2010, Decimal('221328.1420'), 113000, 484147),
 (2020, Decimal('197155.7467'), 97393, 613027)]

Discovery: Average song durations rose slightly from the 70s to the 90s, where they peaked, and then have fallen ever since, with them now being down by over a minute in the 2020s so far, when compared to the 90s peak.  There also seems to me some outlier top hits that are really long for their decade, especially in the 20th century.

Insight: This proves the theory that top hits are getting shorter and shows that modern consumers show preference toward shorter songs.  As such, current songs should be made to be relatively short if they want better chances of being a top hit.

# PRIMARY QUESTION SQL QUERY

Primary Question: What are the characteristics of a hit song?

Business Justification: Knowing the characteristics of the average hit song helps to know what artists to invest in - the ones whose songs have similar characteristics to that of a top hit.

SQL Features used: GROUP BY, JOIN, VIEW, Subquery, Window Function

a) First create a view combining the track, track_attributes, and track_details tables into 1 large table.

In [18]:
%%sql

CREATE VIEW top_songs_view AS (
	SELECT
		t.track_id,
		t.chart_year,
		t.track_name,
		t.artist,
		t.album,
		td.track_key,
		td.tempo,
		td.duration_ms,
		td.time_signature,
		ta.danceability,
		ta.energy,
		ta.loudness,
		ta.speechiness,
		ta.acousticness,
		ta.instrumentalness,
		ta.liveness,
		ta.valence
	FROM track t
	JOIN track_details td 
		ON t.track_id = td.track_id AND t.chart_year = td.chart_year
	JOIN track_attributes ta
		ON t.track_id = ta.track_id AND t.chart_year = ta.chart_year
	ORDER BY t.chart_year
);

 * mysql://admin:***@isba-dev-01.ciyo3znowdxs.us-east-1.rds.amazonaws.com/final_project
(MySQLdb._exceptions.OperationalError) (1050, "Table 'top_songs_view' already exists")
[SQL: CREATE VIEW top_songs_view AS (
	SELECT
		t.track_id,
		t.chart_year,
		t.track_name,
		t.artist,
		t.album,
		td.track_key,
		td.tempo,
		td.duration_ms,
		td.time_signature,
		ta.danceability,
		ta.energy,
		ta.loudness,
		ta.speechiness,
		ta.acousticness,
		ta.instrumentalness,
		ta.liveness,
		ta.valence
	FROM track t
	JOIN track_details td 
		ON t.track_id = td.track_id AND t.chart_year = td.chart_year
	JOIN track_attributes ta
		ON t.track_id = ta.track_id AND t.chart_year = ta.chart_year
	ORDER BY t.chart_year
);]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


b) Find the characteristics of the average top hit.

In [19]:
%%sql

SELECT
	(
		SELECT track_key
        FROM top_songs_view
        GROUP BY track_key
        ORDER BY COUNT(*) DESC
        LIMIT 1
    ) AS most_common_track_key,
    AVG(tempo) AS avg_tempo,
    AVG(duration_ms) AS avg_duration_ms,
    (
		SELECT time_signature
		FROM top_songs_view
		GROUP BY time_signature
		ORDER BY COUNT(*) DESC
		LIMIT 1
  	) AS most_common_time_signature,
    AVG(danceability) AS avg_danceability,
    AVG(energy) AS avg_energy,
    AVG(loudness) AS avg_loudness,
    AVG(speechiness) AS avg_speechiness,
    AVG(acousticness) AS avg_acousticness,
    AVG(instrumentalness) AS avg_instrumentalness,
    AVG(liveness) AS avg_liveness,
    AVG(valence) AS avg_valence
FROM top_songs_view;

 * mysql://admin:***@isba-dev-01.ciyo3znowdxs.us-east-1.rds.amazonaws.com/final_project
1 rows affected.


[('C', 119.88951981879903, Decimal('241572.7488'), 4, 0.63501528878822, 0.6618443563608891, -7.806686107965287, 0.07448057757644408, 0.20013455896564755, 0.02822715503020014, 0.17394705549263859, 0.5913461494903739)]

Insight: The values listed in the results chart show the general characteristics of a hit song (from the 1970s through 2022).

Actionable Reccomendation: Invest in artists whose song library's characteristics closely mirror these average characteristics of a hit song.

# SECONDARY QUESTION #1 SQL QUERY

Secondary Question 1: How have the characteristics of a hit changed by the decade?

Business Justification: To get more accurate characteristics for modern top hits and to see the direction that consumer preferences in top hits has been moving.

SQL Features used: GROUP BY, CASE, VIEW, Subquery, Window Function

In [20]:
%%sql

SELECT
    FLOOR(chart_year / 10) * 10 AS decade,
    (
        SELECT track_key
        FROM top_songs_view
        WHERE chart_year >= decade AND chart_year < decade + 10
        GROUP BY track_key
        ORDER BY COUNT(*) DESC
        LIMIT 1
    ) AS most_common_track_key,
    AVG(tempo) AS avg_tempo,
    AVG(duration_ms) AS avg_duration_ms,
    (
        SELECT time_signature
        FROM top_songs_view
        WHERE chart_year >= decade AND chart_year < decade + 10
        GROUP BY time_signature
        ORDER BY COUNT(*) DESC
        LIMIT 1
    ) AS most_common_time_signature,
    AVG(danceability) AS avg_danceability,
    AVG(energy) AS avg_energy,
    AVG(loudness) AS avg_loudness,
    AVG(speechiness) AS avg_speechiness,
    AVG(acousticness) AS avg_acousticness,
    AVG(instrumentalness) AS avg_instrumentalness,
    AVG(liveness) AS avg_liveness,
    AVG(valence) AS avg_valence,
    (	COUNT(
	    	CASE
		    	WHEN tempo >= 120
		    	THEN 1
		    	ELSE NULL
		    END)
		/
		COUNT(*)
	) AS percent_fast_songs
FROM top_songs_view
GROUP BY decade
ORDER BY decade;

 * mysql://admin:***@isba-dev-01.ciyo3znowdxs.us-east-1.rds.amazonaws.com/final_project
6 rows affected.


[(1970, 'A', 118.92149500000004, Decimal('239874.0250'), 4, 0.576191, 0.5895440000000016, -10.082174999999985, 0.051627899999999956, 0.3241795260100006, 0.03935840128999995, 0.1771675000000001, 0.6577794999999993, Decimal('0.4830')),
 (1980, 'C', 121.446647, Decimal('259409.5730'), 4, 0.6353920000000005, 0.6516105000000002, -9.683818999999994, 0.047810599999999995, 0.21206127165999983, 0.036974889680000025, 0.16555819999999988, 0.6544603000000007, Decimal('0.4790')),
 (1990, 'C', 118.46770070070077, Decimal('261809.5876'), 4, 0.6304544544544548, 0.6782312312312323, -8.331006006005998, 0.06726686686686685, 0.1590301715015016, 0.04481014054054049, 0.178754154154154, 0.5832598598598603, Decimal('0.4520')),
 (2000, 'C#', 119.88563463463447, Decimal('238784.9660'), 4, 0.6553263263263257, 0.7196741741741743, -5.607687687687692, 0.10109139139139149, 0.1373592953953953, 0.016290925935935938, 0.17761641641641623, 0.5842516516516516, Decimal('0.4760')),
 (2010, 'C#', 120.17245500000004, Decimal('221328.1420'), 4, 0.660719, 0.6814584000000001, -5.757431999999996, 0.0955256, 0.16006014839999994, 0.010233225470000008, 0.17207820000000018, 0.5004058999999998, Decimal('0.5170')),
 (2020, 'C#', 121.73032333333336, Decimal('197155.7467'), 4, 0.6917133333333341, 0.624436666666667, -6.372139999999993, 0.1048136666666667, 0.2263962330000001, 0.006469951666666665, 0.16917800000000002, 0.5132073333333332, Decimal('0.4900'))]

Insight: Trends include that modern hits (from the 2020s) have recently been getting shorter, more dancable, less energetic, more speechy, more acoustic, and less instrumental.  Top hits also generally having high loudness compared to last century, but it has been very gradually dropping since the 2000s.  And tracks have been getting progressively more negative over time since the 70s (as indicated by the falling valence characteristics).

Actionable Reccomendation: Encourage artists to make songs a bit shorter, more danceable, and less instrumental.  Ecourage mastering engineers to master songs not quite as loud, allowing for more dynamics.  Doing so should appeal to a larger audience and drive steams/revenue.


# SECONDARY QUESTION #2 SQL QUERY

Secondary Question 2: What characteristics do the 100 artists with the most Top Hits have and how do they compare to the average characteristics of all top hits?

Business Justification: To see if there are noticable differences in what makes a popular artist and a very popular artist.  Knowing this can help to invest in only the artists with the absolute highest chances of having top hits.

SQL Features used: GROUP BY, VIEW, CTE, Subquery, Window Function

In [22]:
%%sql

WITH top_100_artists AS (
	SELECT 
	    artist,
	    COUNT(DISTINCT track_id) AS number_of_top_hits,
	    (
	        SELECT track_key
	        FROM top_songs_view
	        WHERE artist = t.artist
	        GROUP BY track_key
	        ORDER BY COUNT(*) DESC
	        LIMIT 1
	    ) AS most_common_track_key,
	    AVG(tempo) AS avg_tempo,
	    AVG(duration_ms) AS avg_duration_ms,
	    (
	        SELECT time_signature
	        FROM top_songs_view
	        WHERE artist = t.artist
	        GROUP BY time_signature
	        ORDER BY COUNT(*) DESC
	        LIMIT 1
	    ) AS most_common_time_signature,
	    AVG(danceability) AS avg_danceability,
	    AVG(energy) AS avg_energy,
	    AVG(loudness) AS avg_loudness,
	    AVG(speechiness) AS avg_speechiness,
	    AVG(acousticness) AS avg_acousticness,
	    AVG(instrumentalness) AS avg_instrumentalness,
	    AVG(liveness) AS avg_liveness,
	    AVG(valence) AS avg_valence
	FROM top_songs_view t
	GROUP BY artist
	ORDER BY number_of_top_hits DESC
	LIMIT 100
	)
SELECT
	"top_100_artists_hits" AS hit_type,
	(
		SELECT most_common_track_key
        FROM top_100_artists
        GROUP BY most_common_track_key
        ORDER BY COUNT(*) DESC
        LIMIT 1
    ) AS most_common_track_key,
    AVG(avg_tempo) AS avg_tempo,
    AVG(avg_duration_ms) AS avg_duration_ms,
    (
		SELECT most_common_time_signature
		FROM top_100_artists
		GROUP BY most_common_time_signature
		ORDER BY COUNT(*) DESC
		LIMIT 1
  	) AS most_common_time_signature,
    AVG(avg_danceability) AS avg_danceability,
    AVG(avg_energy) AS avg_energy,
    AVG(avg_loudness) AS avg_loudness,
    AVG(avg_speechiness) AS avg_speechiness,
    AVG(avg_acousticness) AS avg_acousticness,
    AVG(avg_instrumentalness) AS avg_instrumentalness,
    AVG(avg_liveness) AS avg_liveness,
    AVG(avg_valence) AS avg_valence
FROM top_100_artists
UNION ALL
SELECT
	"all_top_hits" AS hit_type,
	(
		SELECT track_key
        FROM top_songs_view
        GROUP BY track_key
        ORDER BY COUNT(*) DESC
        LIMIT 1
    ) AS most_common_track_key,
    AVG(tempo) AS avg_tempo,
    AVG(duration_ms) AS avg_duration_ms,
    (
		SELECT time_signature
		FROM top_songs_view
		GROUP BY time_signature
		ORDER BY COUNT(*) DESC
		LIMIT 1
  	) AS most_common_time_signature,
    AVG(danceability) AS avg_danceability,
    AVG(energy) AS avg_energy,
    AVG(loudness) AS avg_loudness,
    AVG(speechiness) AS avg_speechiness,
    AVG(acousticness) AS avg_acousticness,
    AVG(instrumentalness) AS avg_instrumentalness,
    AVG(liveness) AS avg_liveness,
    AVG(valence) AS avg_valence
FROM top_songs_view;

 * mysql://admin:***@isba-dev-01.ciyo3znowdxs.us-east-1.rds.amazonaws.com/final_project
2 rows affected.


[('top_100_artists_hits', 'C#', 119.25030658241971, Decimal('247378.09933700'), 4, 0.6326439607431261, 0.6673047265974745, -7.2299221363809485, 0.07652232943521513, 0.18782120345618586, 0.01963876265678094, 0.17514787454924188, 0.5596049983595652),
 ('all_top_hits', 'C', 119.88951981879903, Decimal('241572.74877312'), 4, 0.63501528878822, 0.6618443563608891, -7.806686107965287, 0.07448057757644408, 0.20013455896564755, 0.02822715503020014, 0.17394705549263859, 0.5913461494903739)]

Insight: The top hits coming from the top 100 artists with the most top hits generally are slightly longer and slightly louder than the average top hit, but the general characteristics are all so similar that any differences are of negligable size.

Actionable Reccomendation: Beyond finding artists whose song library's characteristics are similar to the average characteristics of hit songs, do not put particular effort in trying to find the absolute best-of-the-best artists with the most potential for hit-making, as those artists' songs' characteristics aren't much different at all from the average top hit.  As such they lack proper indicators for this sort of decision-making.
